### PR TITLE
reading and writing registers

### PR DESCRIPTION
**Pull request by arnabcs17b006**

### CODE REVIEW

File: examples/regs-dump.rs [hunk 1]
```diff
@@ -3,7 +3,7 @@ extern crate env_logger;
 extern crate microvmi;
 
 // traits method can only be used if the trait is in the scope
-use microvmi::api::{Introspectable, Registers};
+use microvmi::api::*;
```
[the remark this answers is not in the document]
Comment (rageagainsthepc) [reply examples/regs-dump.rs:6]: I second that opinion. There are exceptions of course (e.g. `use super::*` for unit tests)

File: Cargo.toml [hunk 2]
```diff
@@ -24,13 +25,14 @@ libc = { version = "0.2.58", optional = true }
...
 xenforeignmemory = { git = "https://github.com/Wenzel/xenforeignmemory", optional = true }
-kvmi = { git = "https://github.com/Wenzel/kvmi", optional = true }
+kvmi = { git = "https://github.com/arnabcs17b006/kvmi", optional = true }
 fdp = { git = "https://github.com/Wenzel/fdp", optional = true }
 winapi = { version = "0.3.8", features = ["tlhelp32", "winnt", "handleapi", "securitybaseapi"], optional = true }
 widestring = { version = "0.4.0", optional = true }
 ntapi = { version = "0.3.3", optional = true }
 vid-sys = { version = "0.3.0", features = ["deprecated-apis"], optional = true }
 cty = "0.2.1"
+clippy = {version = "0.0.188", optional = true}
```
[the remark this answers is not in the document]
Comment (rageagainsthepc) [reply Cargo.toml:35]: I would say absolutely not. Why did you add that in the first place, @arnabcs17b006?

File: src/driver/kvm.rs [hunk 3]
```diff
@@ -3,12 +3,9 @@ use std::error::Error;
...
 use std::vec::Vec;
 
-use kvmi::{KVMi, KVMiCr, KVMiEvent, KVMiEventReply, KVMiEventType, KVMiInterceptType};
+use kvmi::{KVMi, KVMiCr, KVMiEvent, KVMiEventReply, KVMiEventType, KVMiInterceptType, KVMiRegs};
 
-use crate::api::{
-    CrType, DriverType, Event, EventReplyType, EventType, InterceptType, Introspectable, Registers,
-    X86Registers,
-};
+use crate::api::*;
```
[the remark this answers is not in the document]
Comment (arnabcs17b006) [reply src/driver/kvm.rs:8]: Sir all the members of api.rs are being used

**Comment by Wenzel** [comment]
Try to rebase your PR on master, we fixed some issue in the CI, see if it fixes the vbox problem

File: Cargo.toml [hunk 4]
```diff
@@ -36,4 +36,4 @@ cty = "0.2.1"
 env_logger = "0.7.1"
 ctrlc = "3.1.3"
 clap = "2.33.0"
-colored = "1.9.3"
+colored = "1.9.3"
```
Comment (rageagainsthepc) [inline Cargo.toml:39]: You don't need to change anything here but I suggest you install the editorconfig plugin. I have added an editorconfig file to the master branch which will generally make your coding experience a little bit more consistent.

File: src/driver/kvm.rs [hunk 5]
```diff
@@ -85,12 +86,94 @@ impl Introspectable for Kvm {
...
 
+    fn write_registers(&self, vcpu: u16, reg: Registers) -> Result<(), Box<dyn Error>> {
+        let register: X86Registers;
+        match reg {
+            Registers::X86(x86_registers) => {
+                register = x86_registers;
+            }
+        }
+
+        let regs = KVMiRegs {
```
Comment (rageagainsthepc) [inline src/driver/kvm.rs:151]: nitpick: You could also use the From trait here which would reduce the amount of lines. Then you could put everything into the match arm above.

File: src/driver/kvm.rs [hunk 6]
```diff
@@ -85,12 +86,94 @@ impl Introspectable for Kvm {
...
-            fs_base: sregs.fs.base,
+            sysenter_cs: _msrs.entries[0].data,
+            sysenter_esp: _msrs.entries[1].data,
+            sysenter_eip: _msrs.entries[2].data,
+            msr_efer: _msrs.entries[3].data,
+            msr_star: _msrs.entries[4].data,
+            msr_lstar: _msrs.entries[5].data,
+            efer: sregs.efer,
+            apic_base: sregs.apic_base,
+            cs: segment_reg {
```
[the remark this answers is not in the document]
Comment (rageagainsthepc) [reply src/driver/kvm.rs:100]: You can re-export the `kvm_segment` in the kvmi crate in order to make it available here.

[the remark this answers is not in the document]
Comment (arnabcs17b006) [reply src/driver/kvm.rs:100]: I am not clear about what are you saying here

File: src/driver/kvm.rs [hunk 7]
```diff
@@ -87,12 +121,38 @@ impl<T: KVMIntrospectable> Introspectable for Kvm<T> {
...
-            fs_base: sregs.fs.base,
+            sysenter_cs: _msrs.entries[0].data,
+            sysenter_esp: _msrs.entries[1].data,
+            sysenter_eip: _msrs.entries[2].data,
+            msr_efer: _msrs.entries[3].data,
+            msr_star: _msrs.entries[4].data,
+            msr_lstar: _msrs.entries[5].data,
+            efer: sregs.efer,
+            apic_base: sregs.apic_base,
+            cs: SegmentReg::from(sregs.cs),
```
Comment (rageagainsthepc) [inline src/driver/kvm.rs:135]: That looks so much nicer already :+1: But if you implement `From` you get `Into` for free, so that means you can just write:
```suggestion
            cs: sregs.cs.into(),
```

File: src/driver/kvm.rs [hunk 8]
```diff
@@ -20,6 +20,41 @@ pub struct Kvm<T: KVMIntrospectable> {
     vec_events: Vec<Option<KVMiEvent>>,
 }
 
+impl From<kvm_segment> for SegmentReg {
```
Comment (rageagainsthepc) [inline src/driver/kvm.rs:23]: I'd prefer it if you didn't separate the kvm struct from its impl block here.

File: src/driver/kvm.rs [hunk 9]
```diff
@@ -87,12 +121,38 @@ impl<T: KVMIntrospectable> Introspectable for Kvm<T> {
...
+            tr: SegmentReg::from(sregs.tr),
+            ldt: SegmentReg::from(sregs.ldt),
         }))
     }
 
+    fn write_registers(&self, vcpu: u16, reg: Registers) -> Result<(), Box<dyn Error>> {
+        match reg {
+            Registers::X86(x86_registers) => {
+                let regs = kvm_regs::from(x86_registers);
+                self.kvmi.set_registers(vcpu, &regs)?;
```
Comment (rageagainsthepc) [inline src/driver/kvm.rs:150]:
```suggestion
                self.kvmi.set_registers(vcpu, &x86_registers.into())?;
```

File: src/driver/kvm.rs (hunk 8 above)
Comment (arnabcs17b006) [reply src/driver/kvm.rs:23]: Can you please elaborate. I don't quite get it.

File: src/driver/kvm.rs [hunk 10]
```diff
@@ -205,95 +265,9 @@ impl<T: KVMIntrospectable> Drop for Kvm<T> {
             self.kvmi
                 .control_events(vcpu, KVMiInterceptType::Cr, false)
                 .unwrap();
-        }
-    }
-}
-
-#[cfg(test)]
```
Comment (rageagainsthepc) [inline src/driver/kvm.rs:212]: Why are you deleting the tests? 😱

File: src/driver/xen.rs [hunk 11]
```diff
@@ -125,6 +173,10 @@ impl Introspectable for Xen {
         debug!("resume");
         Ok(self.xc.domain_unpause(self.domid)?)
     }
+
```
Comment (rageagainsthepc) [inline src/driver/xen.rs:176]: ?

File: src/driver/kvm.rs (hunk 10 above)
Comment (arnabcs17b006) [reply src/driver/kvm.rs:212]: Sir
The tests are present in the unit_tests branch

Comment (rageagainsthepc) [reply src/driver/kvm.rs:212]: Yes, your new tests. But those are the tests that are already present in the master branch. You should not delete them.

File: src/driver/xen.rs (hunk 11 above)
Comment (rageagainsthepc) [reply src/driver/xen.rs:176]: Please delete this.

File: c_examples/mem-dump.c [hunk 12]
```diff
@@ -47,4 +47,4 @@ int main(int argc, char* argv[]) {
     dump_memory(driver, argv[1]);
     microvmi_destroy(driver);
     return 0;
-}
```
Comment (rageagainsthepc) [inline c_examples/mem-dump.c:50]: Please try to get rid of all the unnecessary whitespace changes.

File: src/driver/kvm.rs [hunk 13]
```diff
@@ -58,15 +93,14 @@ impl<T: KVMIntrospectable> Introspectable for Kvm<T> {
...
-        // No API in KVMi at the moment
-        // fake 512MB
-        let max_addr = 1024 * 1024 * 512;
-        Ok(max_addr)
+        let max_gfn = self.kvmi.get_maximum_gfn()?;
+        Ok(max_gfn << PAGE_SHIFT)
     }
 
     fn read_registers(&self, vcpu: u16) -> Result<Registers, Box<dyn Error>> {
         let (regs, sregs, _msrs) = self.kvmi.get_registers(vcpu)?;
```
Comment (rageagainsthepc) [inline src/driver/kvm.rs:101]:
```suggestion
        let (regs, sregs, msrs) = self.kvmi.get_registers(vcpu)?;
```

File: src/driver/kvm.rs [hunk 14]
```diff
@@ -58,15 +93,14 @@ impl<T: KVMIntrospectable> Introspectable for Kvm<T> {
...
-        Ok(max_addr)
+        let max_gfn = self.kvmi.get_maximum_gfn()?;
+        Ok(max_gfn << PAGE_SHIFT)
     }
 
     fn read_registers(&self, vcpu: u16) -> Result<Registers, Box<dyn Error>> {
-        let (regs, sregs, _msrs) = self.kvmi.get_registers(vcpu)?;
+        let (regs, sregs, msrs) = self.kvmi.get_registers(vcpu)?;
         // TODO: hardcoded for x86 for now
+        let msrs_as_slice = msrs.as_slice();
```
Comment (rageagainsthepc) [inline src/driver/kvm.rs:103]: The comment in the line above belongs to the line below.

File: src/driver/dummy.rs [hunk 15]
```diff
@@ -1,6 +1,5 @@
 use crate::api::Introspectable;
 use std::error::Error;
-
```
Comment (rageagainsthepc) [inline src/driver/dummy.rs:3]: Unnecessary whitespace change.

File: c_examples/regs-dump.c [hunk 16]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (rageagainsthepc) [inline c_examples/regs-dump.c:1]: Unnecessary whitespace change.

File: c_examples/mem-dump.c [hunk 17]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (rageagainsthepc) [inline c_examples/mem-dump.c:1]: Unnecessary whitespace change.

File: c_examples/Makefile [hunk 18]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (rageagainsthepc) [inline c_examples/Makefile:1]: Unnecessary whitespace change.

File: src/api.rs [hunk 19]
```diff
@@ -67,6 +93,11 @@ pub trait Introspectable {
         unimplemented!();
     }
 
+    // write registers
```
Comment (rageagainsthepc) [inline src/api.rs:96]: This comment provides no additional information.

File: src/driver/xen.rs [hunk 20]
```diff
@@ -125,6 +173,9 @@ impl Introspectable for Xen {
         debug!("resume");
         Ok(self.xc.domain_unpause(self.domid)?)
     }
+    /*fn get_driver_type(&self) -> DriverType {
```
Comment (rageagainsthepc) [inline src/driver/xen.rs:176]: Again, please delete the whole range of commented out code.
